### PR TITLE
Add python scripts for automated stat testing and visualisation

### PR DESCRIPTION
**Pull request by heidmic**
These scripts should contain a lot of functionality to do automated stat. testing and visualisations/tables

We want to expand this functionality in the future

### CODE REVIEW

**Comment by heidmic** [comment]
rename folder to hint at what is contained

**Review by RomanSraj — CHANGES_REQUESTED** [review]

**Comment by hubanton** [comment]
Implemented the requested changes and added further comments for the sake of understandability.

File: logging_output_scripts/latex_tabulars.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    for v in datasets.values():
+        res = df[df['Problem'].str.contains(v)]
+        data_res.append(float(res[c]))
+        return data_res
+
+
+# COMPLEXITY TABLES
+def write_complexity():
+    for i in dirs:
+        directory = dirs[i]
```
Comment (RomanSraj) [inline logging_output_scripts/latex_tabulars.py:43]: Why not use `for directory in dirs.values()`? Then you can remove line 43

File: logging_output_scripts/latex_tabulars.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        df = pd.read_csv(f"../{directory}/Results.csv")
+        comp_list = []
+        for j in comp_column:
+            c = comp_column[j]
```
Comment (RomanSraj) [inline logging_output_scripts/latex_tabulars.py:47]: Why not use `for c in comp_column.values()`? Then you can remove line 47

File: logging_output_scripts/latex_tabulars.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+        res = tabulate(comp_list, tablefmt="latex", headers=datasets_short.values())
+        with open(f"Complexity/Complexity-{directory}.txt", "w") as file:
+            file.write(res)
+
+
+# GENOMES
+def write_genomes():
+    for i in dirs:
+        directory = dirs[i]
```
Comment (RomanSraj) [inline logging_output_scripts/latex_tabulars.py:59]: Why not use `for directory in dirs.values()`? Then you can remove line 59

File: logging_output_scripts/latex_tabulars.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        for j in datasets:
+            dataset = datasets[j]
```
Comment (RomanSraj) [inline logging_output_scripts/latex_tabulars.py:61]: Why not use `for dataset in datasets.values()`? Then you can remove line 61

File: logging_output_scripts/latex_tabulars.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+            with open(f'Genomes/{directory}_{dataset}.json', 'r') as f:
+                genomes = json.load(f)
+
+            genomes = list(genomes.values())
+
+            genome_list = []
+            for x in range(len(genomes)):
+                temp = genomes[i]
```
Comment (RomanSraj) [inline logging_output_scripts/latex_tabulars.py:70]: Why are we using `i` here?

File: logging_output_scripts/latex_tabulars.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+                file.write(res)
+
+
+# MSE TABLES (Creates individual tables to be combined into a larger tabular)
+def write_mse():
+    row = []
+    for v in datasets.values():
+        temp = []
+        for i in dirs:
+            directory = dirs[i]
```
Comment (RomanSraj) [inline logging_output_scripts/latex_tabulars.py:85]: Why not use `for directory in dirs.values()`? Then you can remove line 85

File: logging_output_scripts/latex_tabulars.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    table_5 = tabulate(row[4], tablefmt="latex", headers=headers)
+
+    with open(f"MSE/MSE.txt", "w") as file:
+        file.write(table_1 + "\n\n" + table_2 + "\n\n" + table_3 + "\n\n" + table_4 + "\n\n" + table_5)
+
+
+# CONVERGENCE TABLES
+def write_convergence():
+    for i in dirs:
+        directory = dirs[i]
```
Comment (RomanSraj) [inline logging_output_scripts/latex_tabulars.py:105]: Why not use `for directory in dirs.values()`? Then you can remove line 105

File: logging_output_scripts/latex_tabulars.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        df = pd.read_csv(f"../{directory}/Results.csv")
+        comp_list = []
+        for j in thresh_column:
+            c = thresh_column[j]
```
Comment (RomanSraj) [inline logging_output_scripts/latex_tabulars.py:109]: Why not use `for c in thresh_column.values()`? Then you can remove line 109

File: logging_output_scripts/latex_tabulars.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+            data_res = load_problem_columns(df, c)
+            if j % 3 == 0:
+                threshold = int(j / 3)
+                comp_list.append(
+                    (f'Threshold = {threshold}', thresh_column_short[i], data_res[0], data_res[1],
+                     data_res[2], data_res[3], data_res[4]))
+            else:
+                comp_list.append((" ", thresh_column_short[i], data_res[0],
+                                  data_res[1], data_res[2], data_res[3], data_res[4]))
```
Comment (RomanSraj) [inline logging_output_scripts/latex_tabulars.py:118]: Why do we need this `if/else`? 
If we need it, you can write it like this to make it more readable:
```
if j % 3 == 0:
    threshold = int(j / 3)
    threshold_string = f'Threshold = {threshold}'

comp_list.append((threshold_string, thresh_column_short[i], data_res[0],
                                  data_res[1], data_res[2], data_res[3], data_res[4]))
```

File: logging_output_scripts/latex_tabulars.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+        new_header.insert(0, " ")
+        res = tabulate(comp_list, tablefmt="latex", headers=new_header)
+        with open(f"Convergence/convergence-{directory}.txt", "w") as file:
+            file.write(res)
+
+
+# Final iter TABLES
+def write_final_iter():
+    for i in dirs:
+        directory = dirs[i]
```
Comment (RomanSraj) [inline logging_output_scripts/latex_tabulars.py:129]: Why not use `for directory in dirs.values()`? Then you can remove line 129

File: logging_output_scripts/latex_tabulars.py [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+        df = pd.read_csv(f"../{directory}/Results.csv")
+        comp_list = []
+        for j in iter_column:
+            c = iter_column[j]
```
Comment (RomanSraj) [inline logging_output_scripts/latex_tabulars.py:133]: Why not use `for c in iter_column.values()`? Then you can remove line 105

File: logging_output_scripts/summary_csv.py [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+# The used representations (or models) on which runs were performed
+dirs = {0: 'OBR', 1: 'UBR', 2: 'CSR', 3: 'MPR'}
+
+for i in dirs:
+    # Head of csv-File
+    s = f"Problem,MEAN_COMP,STD_COMP,MEDIAN_COMP,MIN_COMP,MAX_COMP,MEAN_MSE,STD_MSE," \
+        f"FIN_ITER_MAX,FIN_ITER_MIN,FIN_ITER_MEAN,THRESH_0_MEAN,THRESH_0_MIN,THRESH_0_MAX," \
+        f"THRESH_1_MEAN,THRESH_1_MIN,THRESH_1_MAX,THRESH_2_MEAN,THRESH_2_MIN,THRESH_2_MAX"
+    # Current working directory
+    directory = dirs[i]
```
Comment (RomanSraj) [inline logging_output_scripts/summary_csv.py:20]: Why not use `for directory in dirs.values()`? Then you can remove line 20

File: logging_output_scripts/summary_csv.py [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+
+    for j in datasets:
+        problem = datasets[j]
```
Comment (RomanSraj) [inline logging_output_scripts/summary_csv.py:23]: Why not use `for problem in datasets.values()`? Then you can remove line 23

File: logging_output_scripts/summary_csv.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+        mean_t2 = thresh_2.mean()
+        max_t2 = thresh_2.max()
+        min_t2 = thresh_2.min()
+        std_t2 = thresh_2.std()
+
+        # Appends row of results for each Dataset
+        s += f"\n\n{problem},{mean_comp},{std_comp},{median_comp},{min_comp},\
+        {max_comp},{mean_mse},{std_mse},{max_final_iter},{min_final_iter},\
+        {mean_final_iter},{mean_t0},{min_t0},{max_t0},{mean_t1},{min_t1},\
+        {max_t1},{mean_t2},{min_t2},{max_t2}"
```
Comment (RomanSraj) [inline logging_output_scripts/summary_csv.py:72]: Wouldn't it make more sense to directly add the strings together:

```
s += problem
s += "," + elitist_complexity.min() 
s += "," + elitist_complexity.max()
s += "," + elitist_complexity.mean()
...
```

File: logging_output_scripts/violin_plots.py [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+
+dirs = {0: 'OBR', 1: 'UBR', 2: 'CSR', 3: 'MPR'}
+
+datasets = {0: 'parkinson_total', 1: 'protein_structure', 2: 'airfoil_self_noise',
+            3: 'concrete_strength', 4: 'combined_cycle_power_plant'}
+
+dict_list = {}
+
+for i in datasets:
+    problem = datasets[i]
```
Comment (RomanSraj) [inline logging_output_scripts/violin_plots.py:18]: Why not use `for problem in datasets.values()`? Then you can remove line 18

File: logging_output_scripts/violin_plots.py [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+        df = pd.read_csv(f"../{directory}/{problem}.csv")
+        fold_df = df[df['Name'].str.contains('fold')]
+        name = []
+        for x in range(fold_df.shape[0]):
+            name.append(directory)
+        # Adds additional column for plotting
+        if j == 0:
+            res_var = fold_df.assign(Used_Representation=name)
+        else:
+            res_var = pd.concat([res_var, fold_df.assign(Used_Representation=name)])
```
Comment (RomanSraj) [inline logging_output_scripts/violin_plots.py:31]: Use a flag for the first iteration

**Review by RomanSraj — CHANGES_REQUESTED** [review]
Most of the change requests are in regard to removing unnecessary lines of code. Please also consider using more descriptive variable names. `c`, `v`, `a` etc make the code harder to read than necessary

File: logging_output_scripts/latex_tabulars.py (hunk 2 above)
Comment (hubanton) [reply logging_output_scripts/latex_tabulars.py:47]: The index is needed to access the other directory containing name for the actual latex tabular. Switched to iterating over both dictionaries simultaneously using zip

File: logging_output_scripts/latex_tabulars.py (hunk 8 above)
Comment (hubanton) [reply logging_output_scripts/latex_tabulars.py:109]: Index is needed to format the tabular properly

File: logging_output_scripts/latex_tabulars.py (hunk 11 above)
Comment (hubanton) [reply logging_output_scripts/latex_tabulars.py:133]: Also switched to using zip, as index was also needed for other dictionary

**Comment by hubanton** [comment]
Implemented all proposed changes, reduced code to only cover the most basic metrics, added automatic folder creating. 

The code now works with the logging settings as found in the main-branch and can be easily extended / adapted. 
Only need to set up csv-Files in respective folders and then all scripts run as extended.

**Pull request merged**